src/test/java/com/sevendragons/practice/leetcode/medium/courseschedule/Solution.java

```diff
@@ -0,0 +1,61 @@
+package com.sevendragons.practice.leetcode.medium.coursesschedule;
+
+import java.util.*;
+
+public class Solution {
+
+    //numCourses : id of course (ex: 3)
+    //prerequesite: depencencies of course (ex: [[3,2],[2,1],[1,2], [3,1]])
+    public boolean canFinish(int numberOfCourse, int[][] prerequisites) {
+
+        // Init graph
+        Map<Integer, Set<Integer>> prerequisitePerCourse = new HashMap<>();
+        for (int[] pair : prerequisites) {
+            if (prerequisitePerCourse.get(pair[0]) == null) {
+                prerequisitePerCourse.put(pair[0], new HashSet<Integer>());
+            }
+            prerequisitePerCourse.get(pair[0]).add(pair[1]);
+        }
+
+        // Detect cycle
+        return !isCyclic(numberOfCourse, prerequisitePerCourse);
+    }
+
+    private static boolean isCyclic(int numberOfCourse, Map<Integer, Set<Integer>> prerequisitePerCourse) {
+
+        for (int course = 0; course < numberOfCourse; course++) {
+            Set<Integer> targetCourses = new HashSet<Integer>();
+            targetCourses.add(course);
+            if (isCourseContainedInPrerequisites(prerequisitePerCourse, targetCourses, course) == true) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean isCourseContainedInPrerequisites(Map<Integer, Set<Integer>> prerequisitePerCourse, Set<Integer> targetCourses, int prerequisiteToCheck) {
+
+        for (int course : targetCourses) {
```

Comment:
This loop is fishy. Consider what happens in the next loop after this one: it adds elements in `targetCourses`, and it makes a recursive call -> the content of `targetCourses` keeps growing as you recurse deeper. And in every recursion, there is loop, re-checking elements it already checked in previous calls. In other words, the deeper you recurse, the more wasted operations in this loop. Something's wrong here.